hibernate-core/src/main/java/org/hibernate/type/descriptor/java/spi/EntityJavaType.java

```diff
@@ -44,7 +46,22 @@ public boolean areEqual(T one, T another) {
 
 	@Override
 	public boolean isInstance(Object value) {
-		return getJavaTypeClass().isAssignableFrom( Hibernate.getClassLazy( value ) );
+		final LazyInitializer lazyInitializer = extractLazyInitializer( value );
+		final Class<T> javaTypeClass = getJavaTypeClass();
+		if ( lazyInitializer != null ) {
+			if ( lazyInitializer.isUninitialized() ) {
+				// Try to avoid initializing the proxy, in case of inheritance this is not always possible
+				final Class<?> persistentClass = lazyInitializer.getPersistentClass();
+				return javaTypeClass.isAssignableFrom( persistentClass )
+						|| ( persistentClass.isAssignableFrom( javaTypeClass ) && javaTypeClass.isAssignableFrom( lazyInitializer.getImplementationClass() ) );
+			}
+			else {
+				return javaTypeClass.isAssignableFrom( lazyInitializer.getImplementationClass() );
+			}
```

Comment:
Actually, you can always first try to use the persistent class.

```suggestion
			return javaTypeClass.isAssignableFrom( lazyInitializer.getPersistentClass() ) || javaTypeClass.isAssignableFrom( lazyInitializer.getImplementationClass() );
```

Reply:
i may be wrong but in case of an unitialized proxy checking
```
( persistentClass.isAssignableFrom( javaTypeClass ) && javaTypeClass.isAssignableFrom( lazyInitializer.getImplementationClass() ) );
```
can avoid to initialize the proxy in few cases